sql-plugin/src/main/scala/com/nvidia/spark/rapids/profiler.scala

```diff
@@ -246,13 +272,18 @@ object ProfilerOnExecutor extends Logging {
       val (completedJobs, completedStages, allDone) =
         w.pluginCtx.ask(ProfileJobStageQueryMsg(jobs, stages))
           .asInstanceOf[(Array[Int], Array[Int], Boolean)]
-      if (completedJobs.nonEmpty || completedStages.nonEmpty) {
-        synchronized {
-          completedJobs.foreach(activeJobs.remove)
-          completedStages.foreach(activeStages.remove)
-          if (activeJobs.isEmpty && activeStages.isEmpty) {
+      synchronized {
+        completedJobs.foreach(activeJobs.remove)
+        completedStages.foreach(activeStages.remove)
+        completedStages.foreach(stageTaskCount.remove)
+        lazy val allActiveTaskLimitStagesDone = activeStages.size > 0 && 
+            activeStages.forall(!stageTaskCount.contains(_))
+        if (activeJobs.isEmpty) {
+          if (activeStages.isEmpty) {
             disable()
             stopPollingDriver()
+          } else if (allActiveTaskLimitStagesDone) {
+            disable()
           }
         }
```

Comment:
I don't think this logic is correct.  Take the simple case where we're just profiling a single stage and only the first task of that stage.  When the first task of that stage appears, we'll populate activeStages and stageTaskCount.  Then when the first task completes, we'll remove the stage from activeStages, making it empty.  Later this will be called on a timer, but activeStages is already empty, so we're just going to disable and stop polling.  The stage isn't actually complete yet, so we leaked entries in stageTaskCount.  As long as we are tracking task counts for stages, we need to keep polling the driver.  I think it needs to be more like this:
```suggestion
        if (activeJobs.isEmpty && activeStages.isEmpty) {
          disable()
          if (stageTaskCount.isEmpty) {
            stopPollingDriver()
          }
        }
```

Reply:
> we'll remove the stage from activeStages

The current logic is like:  When the first task completes, we'll remove the stage from `stageTaskCount`, making it empty. But it will still in `activeStages` because we need to remove it at the `onStageCompleted` listener function which will always clean the `stageTaskCount` to avoid leaks.

So, in `updateActiveFromDriver`, we want to `disable` when all active stages have finished their task limit profiling, and `stopPollingDriver` when there are no active stages.

Reply:
> The current logic is like: When the first task completes, we'll remove the stage from stageTaskCount

Are you sure?  See the task completion logic here: https://github.com/NVIDIA/spark-rapids/pull/11708/files#diff-93c022f5f1c4c61dc28b1e9b1653d876d517a042bfad4a465125eecd2c9f8833R140-R145  It's removing from activeStages, not stageTaskCount.  We can't remove from stageTaskCount when a task completes or we'll mistakenly think that the count went back to zero.

The only place where we remove a stage from stageTaskCount is when the driver polling response says the stage has completed.  That won't necessarily happen after the first task completes because there can be many tasks in that stage.

Reply:
oh sorry you are 100% right, updated the code.